### PR TITLE
Faster __vsc_escape_value for bash

### LINKED ISSUE #190255: Bash Shell integration takes a long time to parse long commands
Type: <b>Performance Issue</b>

In a bash terminal with Shell Integration, execute the following command, which is 5k characters long:

```shell
time echo 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam ornare rutrum sem at tempor. Phasellus at est erat. Quisque consequat nisi ac arcu finibus rhoncus. Praesent accumsan purus lectus, in sollicitudin dui posuere vel. Integer dapibus aliquam erat, eu malesuada velit venenatis fermentum. Integer eu lacinia purus. Duis at porttitor nisi. Nunc varius vel nibh quis porta.  Fusce auctor elit a mollis tempus. Vestibulum ullamcorper justo sed mollis rutrum. Etiam rutrum leo ut egestas scelerisque. In augue elit, feugiat ac velit eu, imperdiet pretium nisl. Praesent venenatis sem nec nunc tempus dictum. Aenean euismod velit vel purus euismod suscipit. Aenean vitae diam et nisl tincidunt egestas. Integer pulvinar imperdiet ligula ut tristique. Aenean quis massa est. Quisque consequat fringilla pretium. Aliquam erat volutpat.  Curabitur malesuada a odio id tempus. Cras ut tempor ipsum, et tincidunt leo. Sed sollicitudin tellus sed mi vulputate mattis. Nunc pulvinar, quam et pellentesque laoreet, magna mi lacinia purus, sit amet auctor ligula neque non augue. Donec malesuada semper nisi. Proin posuere at ante eu lobortis. Proin vel ultricies justo, in auctor nulla. Sed lacinia vestibulum iaculis. Sed eu volutpat elit, et pellentesque orci. In varius dolor sit amet tellus auctor tincidunt. Ut commodo feugiat ornare. Proin dignissim laoreet hendrerit. Ut quis sapien ex. Mauris odio turpis, aliquam id rhoncus vel, vehicula eget turpis. Mauris eu erat ac ex commodo ultrices. Ut id mollis justo.  Integer pharetra, nibh sed mattis congue, dolor orci ultricies arcu, a efficitur nisi nunc non leo. Aenean sollicitudin enim in laoreet interdum. Proin egestas massa sed felis lobortis, sed tincidunt mauris scelerisque. Donec ut efficitur nulla. Vivamus consectetur arcu eu odio pretium eleifend. In elit ex, scelerisque vitae venenatis in, mollis id dui. Ut ullamcorper interdum fringilla. Maecenas sit amet nunc eros. Aliquam dapibus sem ac pulvinar elementum. Vestibulum sagittis risus vitae venenatis auctor. Fusce vulputate maximus lacus, eget mollis est efficitur eu. Donec eu hendrerit augue, in convallis erat. Curabitur convallis sollicitudin ex eget bibendum. Vestibulum malesuada velit sit amet risus pretium, eget consectetur arcu dapibus. Morbi placerat varius felis. Donec in neque a tellus vehicula lobortis a laoreet turpis.  Nullam ultrices sagittis porta. Nam elementum tincidunt libero. Nunc consequat elit laoreet sollicitudin cursus. Vestibulum egestas nulla justo. Nulla at tempus nisi. Duis ultrices mollis quam, ac vulputate sapien. Aenean euismod nisl libero, at consequat quam rhoncus sit amet. Aliquam erat volutpat. Etiam imperdiet, magna et viverra fringilla, mauris sem sollicitudin erat, sed tempor ipsum ipsum blandit augue. Suspendisse laoreet condimentum euismod. Vivamus non velit neque. Nunc vel leo sed metus congue sollicitudin. Suspendisse non augue in neque pellentesque fermentum id ac justo.  Cras maximus viverra diam eu viverra. Ut accumsan laoreet risus, id ornare velit condimentum ut. Sed orci diam, volutpat ac hendrerit porta, ornare in massa. Sed auctor, tellus et maximus gravida, metus ipsum luctus est, nec interdum ex enim at ipsum. Duis cursus tortor lectus, nec mattis enim pharetra et. Sed fermentum facilisis lacus ac auctor. Duis convallis ipsum nec velit pulvinar placerat. Integer auctor elit rutrum placerat sodales. Vivamus lacinia augue rhoncus nulla iaculis consectetur.  Fusce vehicula laoreet nisi, non scelerisque sem cursus a. Suspendisse neque ipsum, faucibus nec elit vel, lobortis egestas diam. In a tortor euismod, pellentesque augue et, mollis urna. In hac habitasse platea dictumst. Quisque lobortis urna tortor, et iaculis diam porttitor id. Pellentesque eget eros quis dolor consequat elementum eu vitae felis. Proin nec sodales erat, vitae iaculis purus.  Ut condimentum pellentesque purus, ut eleifend purus accumsan sed. Nam maximus diam ac porttitor consequat. Donec maximus volutpat nunc eu tristique. Morbi consequat volutpat rutrum. Nam tellus massa, euismod vitae consequat tristique, interdum nec mauris. Nulla facilisi. Nunc rhoncus massa nec nibh scelerisque, in laoreet dui varius. Etiam accumsan, diam posuere pretium aliquam, sapien nulla rutrum risus, sed dictum felis elit hendrerit ante. Suspendisse vel risus tortor. Pellentesque semper dui id lorem ornare cursus. Nam porta elit non massa vulputate, non malesuada tortor congue. Maecenas fermentum vulputate lacus, a interdum nisl lacinia in. Suspendisse id sodales lacus. Donec commodo elit porta egestas suscipit. Proin vulputate tellus sit amet velit tempus, eget dapibus nunc dapibus. Vivamus hendrerit, lorem vel dignissim fringilla, purus nisl eleifend nunc, eget iaculis felis augue in orci.  Suspendisse tincidunt vitae est sed elementum. Praesent tempor viverra risus non scelerisque. Duis egestas eros ut ante sodales, vel ullamcorper elit elementum. Maecenas id justo ante. Sed facilisis eleifend mi et posuere. Aenean sollicitudin lorem eros.'
```

On both my machines (Linux native and Windows + RemoteSSH), this commande takes 1s to execute.

Make it 10k characters and it takes 4s. At 15k, it takes 8s.

I'm using bash version 4.4.23 (the one shipped on SLES15 SP4).

It appears to be related to the Bash Shell Integration, specifically the `__vsc_escape_value` function located [here](https://github.com/microsoft/vscode/blob/6e2cace51137529800248f45dce5c3ac4f2a0570/src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh#L100).

This can even be reproduced outside of VSCode, in plain bash. Just define this function in a shell and execute it with a long argument.

As a workaround, it is possible to disable Shell Integration in the settings and open new terminals.

With a newer bash 5.2 built by myself, performances are better, but still non-negligible: 5k: 0.5s, 10k: 1.5s, 15k: 3.5s.\
Note: I get similar numbers with a bash 4.4 compiled by myself, so better performances with newer bash might just be related to different compilation options/optimizations.

---

VS Code version: Code 1.81.1 (6c3e3dba23e8fadc360aed75ce363ba185c49794, 2023-08-09T22:18:39.991Z)
OS version: Linux x64 5.14.21-150400.24.28-default
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz (40 x 2200)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|4, 4, 3|
|Memory (System)|62.23GB (46.36GB free)|
|Process Argv|--crash-reporter-id 80cd3c8b-00ff-4b4a-b16f-38172cad0af1|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|undefined|
|XDG_CURRENT_DESKTOP|undefined|
|XDG_SESSION_DESKTOP|undefined|
|XDG_SESSION_TYPE|tty|
</details><details>
<summary>Process Info</summary>

```
CPU %	Mem MB	   PID	Process
    0	   127	3119455	code main
    0	     0	3119461	   zygote
    0	   191	3119635	     gpu-process
    0	     0	3119735	       broker
    0	     0	3119462	   zygote
    0	     0	3119464	     zygote
    1	   191	3119691	window [1] (Settings - Visual Studio Code)
    0	    64	3127076	window [2] (Issue Reporter)
    0	     0	3119642	   utility-network-service
    0	    64	3119919	shared-process
    0	     0	3127527	     /usr/bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
    0	    64	3119920	fileWatcher [1]
    0	    64	3119988	ptyHost
    0	     0	3124823	     /usr/bin/bash
    0	    64	3119989	extensionHost [1]
```

</details>
<details>
<summary>Workspace Info</summary>

```
;
```

</details>
<details><summary>Extensions (1)</summary>

Extension|Author (truncated)|Version
---|---|---
remote-ssh|ms-|0.90.1

</details>

### PR DESCRIPTION
**Pull request by hsfzxjy**
Fix #190255 

This PR presents two faster escaping functions as the replacement of the original slower one.

The first one escapes each byte 0xab into '\xab', which is most scalable and has promising runtime
efficiency, except that it relies on external commands `od` and `tr`.

The second one utilizes bash builtin string substitution, which is much faster and has zero dependency, except that it escapes only `'\\' -> '\\\\'` and `';' -> '\x3b'`, which scales up badly for more patterns.

Currently it defaults to use the second function, but if env `VSCODE_SHELL_FULL_ESCAPE` is set and `od`, `tr` are
available, it will switch to the first one. 

We may discuss more on the choice of the implementation.

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:108]: Should we perhaps select a magic number  for `$1` length and opt for the new approach when it's above that threshold?

Comment (hsfzxjy) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:108]: Sounds reasonable. What about make it 2000 bytes? This is a point where I can observe around 0.1s delay with the old method.

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:108]: On your machine how long does the new method take? 100ms is quite large still

Comment (hsfzxjy) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:108]: Some benchmarks

|Input size (bytes)|string sub(new default)|od+tr(new opt-in)|old|
|:--:|:--:|:--:|:--:|
|2k|1ms|5ms|107ms|
|100k|33ms|48ms|>1 min|

I think with the new method the overhead is tolerable even with quite large input.

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:108]: They both seem very fast. If I understand right, the od+tr method is superior as it escapes more correctly. Since perf is comparable, I think we should default to the od+tr method if they're available without an opt-in variable, then fallback to the more naive method when they're not available.

Comment (hsfzxjy) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:108]: Sure. I will change accordingly. Shall we still set up the magic number?

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:108]: Sure, let's make it 2k

Comment (hsfzxjy) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:108]: BTW, is /dev/null always available for the execution context of `shellIntegration-bash.sh`? The script is using `command -v od &>/dev/null` to detect the availability of od.

**Comment by hsfzxjy** [comment]
The latest push contains these further modification:

 1. `od+tr` becomes the default solution for faster escaping, and if `od` or `tr` is not available, we fallback to use the string substitution solution.
 2. The old `__vsc_escape_value` function is still used when the length of `$1` is below `2000`. Otherwise we switch to the new faster function.

**Comment by hsfzxjy** [comment]
The latest push contains these further modification:

 1. `od+tr` becomes the default solution for faster escaping, and if `od` or `tr` is not available, we fallback to use the string substitution solution.
 2. The old `__vsc_escape_value` function is still used when the length of `$1` is below `2000`. Otherwise we switch to the new faster function.

**Review by Tyriar — DISMISSED** [review]
Thanks @hsfzxjy, this looks good. I'll wait until next month to merge so we get some testing in Insiders before it goes out.

**Review by DonJayamanne — DISMISSED** [review]

**Comment by hsfzxjy** [comment]
Accidental force push dimissed the reviews. Sorry! :pray:

**Review by Tyriar — APPROVED** [review]

**Review by connor4312 — APPROVED** [review]

**Review by Neilblaze — APPROVED** [review]

**Comment by Tyriar** [comment]
This ended up getting reverted due to a hang https://github.com/microsoft/vscode/issues/190255#issuecomment-1737930980

**Pull request merged**
